src/libbson/src/bson/bson-json.c

```diff
@@ -1124,9 +1124,9 @@ _bson_json_read_string (bson_json_reader_t *reader, /* IN */
       } break;
       case BSON_JSON_LF_DECIMAL128: {
          bson_decimal128_t decimal128;
-         bson_decimal128_from_string (val_w_null, &decimal128);
 
-         if (bson->read_state == BSON_JSON_IN_BSON_TYPE) {
+         if (bson_decimal128_from_string (val_w_null, &decimal128) &&
+             bson->read_state == BSON_JSON_IN_BSON_TYPE) {
             bson->bson_type_data.v_decimal128.value = decimal128;
          } else {
             goto BAD_PARSE;
```

Comment:
On failure, `BAD_PARSE` will set the error message as 
```
Invalid input string "0E+xxx", looking for DECIMAL128
```
I think this message is fine as is, but potentially could be more helpful to the user with some changes. I believe it would involve adding some return information from `bson_decimal128_from_string` as to why the parsing fails (because it may not have to do with an overflowing exponent). Open to thoughts.

Reply:
I think it may help to determine the cause of the error with a new function: `bson_decimal128_from_string_w_error`. But I think that is separate from this bug fix and could be considered in a future change.